cpp/ql/src/semmle/code/cpp/ir/implementation/raw/internal/TranslatedStmt.qll

```diff
@@ -147,8 +150,14 @@ class TranslatedReturnValueStmt extends TranslatedReturnStmt, TranslatedVariable
   final override IRVariable getIRVariable() { result = getEnclosingFunction().getReturnVariable() }
 }
 
+/**
+ * The IR translation of a `return` statement that does not return a value. This includes implicit
+ * return statements at the end of `void`-returning functions.
+ */
 class TranslatedReturnVoidStmt extends TranslatedReturnStmt {
-  TranslatedReturnVoidStmt() { not stmt.hasExpr() }
+  TranslatedReturnVoidStmt() {
+    not stmt.hasExpr() and not hasReturnValue(stmt.getEnclosingFunction())
```

Comment:
A return statement in a function with `void` as its return type can have a `void`-typed expression - what happens then?

Reply:
Great question. I’ll add a test case (and probably a fix).

Reply:
Fix and test case added. We didn't handle this correctly before, either. Good catch.